### ISSUE TITLE
callbacks for hookWindowMessage should have their return value sent to OS

### ISSUE THREAD

**Issue by cberescu** [electron/electron#8762]
* Electron version: 1.4.15
* Operating system: windows 10  

### Expected behavior
On some Windows messages a return may be expected.  The return value of the callbacks should be forwarded to the OS

Example : 
WM_QUERYENDSESSION 
From : https://msdn.microsoft.com/en-us/library/windows/desktop/aa376890(v=vs.85).aspx
`if any application returns zero, the session is not ended. `

When windows shuts down it would be usefull to use this to finish app tasks that would have been done normally  on app exit.  

### Actual behavior

 The retun value is not forwarded

**Comment by alespergl** [comment]
@cberescu Can you please provide more context on what is the issue you are experiencing? Is your experience that Electron somehow subverts the default handling of `WM_QUERYENDSESSION`? If you are proposing a feature in relation to `WM_QUERYENDSESSION`, could you be more specific in terms of what tasks you have in mind that you would like to perform in response to that message?
Thanks

**Comment by cberescu** [comment]
What i initially wanted was to stop my app from exiting when windows is shuting down so i will be able to finish some last operations. In the end i resolved this using node-ffi  by overwriting the default window procedure   

I didn't find any  better way in the documentation so i think there is no way to do this or i didn't look good enough.

If there is a way to do this then just ignore what i am saying below. 

My way to resolve this was by using the _ShutdownBlockReasonCreate_ API , i also had to hook the default window procedure so i can parse the window messages. I had to hook the default window procedure becase when it received WM_QUERYENDSESSION it must return 0 so that the _ShutdownBlockReasonCreate_ API will be called.

From my point of view it will be a nice and usefull feature to have this, having the ability to delay the exit for a few seconds until the app is done finishing some last operations.

**Comment by alespergl** [comment]
Thanks for sharing your experience. Your request to process the return value from `hookWindowMessage` callback and pass it on to the system is problematic, because there can be multiple callbacks registered for any message. Obviously there is no way to accumulate all their return values into just one which would be passed on.

Apart from adding a whole feature for supporting logout scenarios, there's one simpler thing which I suppose we could do, which is to handle `WM_QUERYENDSESSION` so that when there's a shutdown block reason in place (installed by `ShutdownBlockReasonCreate`), we would block log out. You would then need to make sure you call `ShutdownBlockReasonDestroy` to allow the system to log the user out.

**Comment by cberescu** [comment]
I think the best option is to add at some point a new feture that will support logout scenarios.  While your solution will work it will still force me to use the API from windows by using ffi or a c++ module.

**Comment by zcbenz** [comment]
Closing since adding return value to `hookWindowMessage` is not going to be added.

**Comment by natetyler** [comment]
@cberescu Could you detail your solution using node-ffi? I have the same need to do some cleanup before allowing windows to exit.

**Comment by cberescu** [comment]
Sure @natetyler ,
First for this to work you need also the `ref` module 


The definition of  win32 API used looks like this in ffi : 

```javascript

var windowsLibraries = {
			user32 : new ffi.Library('user32', {
				'ShutdownBlockReasonCreate':['uint32',['uint32','pointer']],
				'ShutdownBlockReasonDestroy':['bool',['uint32']],
				'SetWindowLongA':['long',['uint32','int','long']],
				'CallWindowProcA':['int32',['long','void *','uint32','uint32','long']]
			}),
			kernel32 : new ffi.Library('kernel32', {
				'GetCurrentThreadId':['uint32',[]],
				'SetProcessShutdownParameters':['bool',['uint32','uint32']]
			})
		}
```


First create the ShutDown Block Procedure : 
```javascript
let mex = windowsLibraries.user32.ShutdownBlockReasonCreate(win.getNativeWindowHandle().readUInt32LE(0),ref.allocCString('Logging out. Please wait...' + '\0', 'ucs2'));
```


Create a new function to replace the original window function(keep the old one so you can call in your own function) : 

```javascript
let handle = windowsLibraries.kernel32.GetCurrentThreadId();
windowManagerHandlers.origProcessHandler = windowsLibraries.user32.SetWindowLongA(win.getNativeWindowHandle().readUInt32LE(0) ,-4, ref.address(windowManagerHandlers.prodHookCallback));
windowsLibraries.kernel32.SetProcessShutdownParameters(514 ,0);
```

The new window function looks like this : 
```javascript
prodHookCallback : ffi.Callback('long', ['void *','uint32', 'uint32','long'],
				function(hwnd,uMsg, wParam,lParam) {
						switch (uMsg) {
							case 17:
								// here you do the exit procedure 
								return 0;
								break;
							default:
								return windowsLibraries.user32.CallWindowProcA(windowManagerHandlers.origProcessHandler,hwnd,uMsg,wParam,lParam);
							break;

						}
				}
			)
```

Do not forget to release the block put on shutdown : 
```javascript
windowsLibraries.user32.ShutdownBlockReasonDestroy(win.getNativeWindowHandle().readUInt32LE(0));
```

I think that is all.

**Comment by rhazari** [comment]
@cberescu Do you have an example electron app where you are handling this Window shutdown procedure?

**Comment by cberescu** [comment]
@rhazari unfortunately not

**Comment by kitadamsadi** [comment]
> ```js
> let handle = windowsLibraries.kernel32.GetCurrentThreadId();
> windowManagerHandlers.origProcessHandler = windowsLibraries.user32.SetWindowLongA(win.getNativeWindowHandle().readUInt32LE(0) ,-4, ref.address(windowManagerHandlers.prodHookCallback));
> windowsLibraries.kernel32.SetProcessShutdownParameters(514 ,0);
> ```

@cberescu  Many thanks for this code!
 
I found that, to work successfully in Electron 8.5.0 (x64), I needed to call SetProcessShutdownParameters() with a priority > the default application process priority of 0x280 to ensure Windows tried to shutdown the Electron main process before the others, e.g.:
SetProcessShutdownParameters(0x290, 0);

I am doing this in a native node module loaded into the main Electron process.

**Comment by savely-krasovsky** [comment]
@cberescu I know it's kinda late for you probably, but !44598 was merged and available in 35.0.0 alpha build. Also I've prepared native package to work with block reason: https://github.com/savely-krasovsky/shutdown-block-reason

So now everyone is able to prevent shutdown and show a reason using Electron.
